File: src/ruby/lib/dependabot/job/models/dependency.rb
# typed: strict
# frozen_string_literal: true

require 'sorbet-runtime'

module Dependabot
  module Job
    module Models
      # Represents a dependency in the context of a Dependabot job.
      class Dependency < T::Struct
        extend T::Sig

        # The name of the dependency.
        const :name, String

        # The current version of the dependency.
        const :version, T.nilable(String), default: nil

        # The requirements for the dependency.
        const :requirements, T::Array[String], default: []

        # The groups this dependency belongs to.
        const :groups, T::Array[String], default: []

        # The source of the dependency.
        const :source, T.nilable(String), default: nil

        # Whether this is a transitive dependency.
        const :is_transitive, T::Boolean, default: false

        sig { params(hash: T::Hash[Symbol, T.untyped]).returns(Dependency) }
        def self.from_hash(hash)
          new(
            name: hash.fetch(:name),
            version: hash[:version],
            requirements: Array(hash[:requirements] || []),
            groups: Array(hash[:groups] || []),
            source: hash[:source],
            is_transitive: !hash[:is_transitive].nil?
          )
        end

        sig { returns(T::Hash[Symbol, T.untyped]) }
        def to_hash
          {
            name: name,
            version: version,
            requirements: requirements,
            groups: groups,
            source: source,
            is_transitive: is_transitive
          }.compact
        end
      end
    end
  end
end

File: src/dotnet/src/Dependabot.Job/Models/Dependency.cs
using System.Collections.Immutable;

namespace Dependabot.Job;

/// <summary>
/// Represents a dependency in the context of a Dependabot job.
/// </summary>
public sealed record Dependency
{
  /// <summary>
  /// Gets or sets the name of the dependency.
  /// </summary>
  public required string Name { get; init; }

  /// <summary>
  /// Gets or sets the current version of the dependency.
  /// </summary>
  public string? Version { get; init; }

  /// <summary>
  /// Gets or sets the requirements for the dependency.
  /// </summary>
  public ImmutableArray<string> Requirements { get; init; } = [];

  /// <summary>
  /// Gets or sets the groups this dependency belongs to.
  /// </summary>
  public ImmutableArray<string> Groups { get; init; } = [];

  /// <summary>
  /// Gets or sets the source of the dependency.
  /// </summary>
  public string? Source { get; init; }

  /// <summary>
  /// Gets or sets a value indicating whether this is a transitive dependency.
  /// </summary>
  public bool IsTransitive { get; init; }
}
